src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -222,18 +222,40 @@ private void UpdateEndpoints()
             bool suppressPathMatching)
         {
             var newPathSegments = routePattern.PathSegments.ToList();
+            int? requiredParameterEndpointRouteOrder = null;
 
             for (var i = 0; i < newPathSegments.Count; i++)
             {
                 // Check if the pattern can be shortened because the remaining parameters are optional
                 //
-                // e.g. Matching pattern {controller=Home}/{action=Index}/{id?} against HomeController.Index
-                // can resolve to the following endpoints:
-                // - /Home/Index/{id?}
-                // - /Home
+                // e.g. Matching pattern {controller=Home}/{action=Index} against HomeController.Index
+                // can resolve to the following endpoints: (ordered by precidence)
                 // - /
-                if (UseDefaultValuePlusRemainingSegmentsOptional(i, action, allDefaults, newPathSegments))
+                // - /Home
+                // - /Home/Index
+                if (UseDefaultValuePlusRemainingSegmentsOptional(
+                    i,
+                    action,
+                    allDefaults,
+                    newPathSegments,
+                    out var hasOptionalOrCatchAllParameter))
                 {
+                    // The route pattern has matching default values AND an optional parameter
+                    // For link generation we need to include a higher priority endpoint with the full
+                    // pattern and the optional parameter as non-optional so it is chosen when the parameter
+                    // route value is used.
+                    // e.g. {controller=Home}/{action=Index}/{id?} will result in: (ordered by precidence)
+                    // - /Home/Index/{id}
+                    // - /
+                    // - /Home
+                    // - /Home/Index
+                    if (hasOptionalOrCatchAllParameter && requiredParameterEndpointRouteOrder == null)
+                    {
+                        // Keep track of the order. The route will be created later with it so it has a higher precidence
+                        requiredParameterEndpointRouteOrder = routeOrder;
+                        routeOrder++;
```

Comment:
I'm anxious about using order for this since this is something we give users to control.

Reply:
All the RouteEndpoints created by MvcEndpointDataSource have an order assigned. A dev doesn't control it.